mesosphere/marathon
# [Do not Merge] Fixes #2875 by implementing AutoScale functionality

MVP for auto scaling applications in Marathon

Things to do:  
- [ ] Fetch and handle SlaveData in its own Actor
- [ ] Guice wiring without Providers
- [ ] Test Cases for the added logic

How to test:
- define an application with autoscaling enabled (see example below)
- start additional slaves and see the app scaled to the number of slaves
- stop one or more slaves and see the app scaled to the number of slaves
- define an application with different constraints (e.g. match slave attributes)
- for faster scale intervals, use `--scale_apps_interval 5000 --auto_scale_force_deployment_timeout 20000`

```json
{
  "id": "/auto/scaled/app",
  "autoScale": {
    "policies": [
      {
        "name": "mesos_agent_count"
      }
    ]
  },
  "cmd": "sleep 1000",
  "cpus": 0.1,
  "mem": 16,
  "constraints": [   [ "hostname",  "UNIQUE"  ]  ],
  "upgradeStrategy": {
    "minimumHealthCapacity": 0.5
  }
}
```
